### PR TITLE
infoblox event collector

### PR DESCRIPTION
**Pull request by esharf**
## Status
- [x] Ready

## Related Issues
fixes: https://jira-hq.paloaltonetworks.local/browse/CIAC-4457

## Description
Added **Infoblox BloxOne ThreatDefense Event Collector**.,

## Minimum version of Cortex XSOAR
- XSIAM

## Does it break backward compatibility?
 - [x] No

## Must have
- [x] Tests
- [x] Documentation

### CODE REVIEW

File: Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefenseEventCollector/InfobloxBloxOneThreatDefenseEventCollector_description.md [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+1. From the SSO portal, click **User Access** from the left navigation panel and then click the **Service API Keys** tab.
+
+2. On the Service API Keys tab, click **Create**.
+
+3. In the Create Service API Keys dialog, complete the following:
+
+  - **Name**: Enter the name of the user API key. Use a name that can identify the purpose of the key.
+
+  - **Service User**: Enter the name of the service users or choose one from the drop-down list. You can associate up to 10 service API keys per service user.
```
Comment (ShirleyDenkberg) [inline Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefenseEventCollector/InfobloxBloxOneThreatDefenseEventCollector_description.md:17]:
```suggestion
  - **Service User**: Enter the name of the service users or choose one from the dropdown list. You can associate up to 10 service API keys per service user.
```

File: Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefenseEventCollector/InfobloxBloxOneThreatDefenseEventCollector.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+def fetch_events_command(client: BloxOneTDEventCollectorClient, params: Dict, last_run: Dict):
+    from_ts = last_run.get('from_ts')
+    if from_ts is None:
+        from_date_time = dateparser.parse(params.get('first_fetch', '1 day'), settings={'TIMEZONE': 'UTC'})
+        if from_date_time is None:
+            raise DemistoException('Invalid date format in "First fetch time interval" parameter')
+
+        from_ts = int(from_date_time.timestamp())
+
+    current_ts = int(datetime.utcnow().timestamp())
```
Comment (bziser) [inline Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefenseEventCollector/InfobloxBloxOneThreatDefenseEventCollector.py:39]: We don't want to take the last event's timestamp to be the `current_ts`?

File: Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefenseEventCollector/InfobloxBloxOneThreatDefenseEventCollector.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+def main():
+    params = demisto.params()
+    client = BloxOneTDEventCollectorClient(
+        api_key=params['credentials']['password'],
+        verify=not argToBoolean(params.get('insecure', False)),
+        proxy=argToBoolean(params.get('proxy', False))
+    )
+
+    command = demisto.command()
+    results: Optional[CommandResults | str] = None
```
Comment (bziser) [inline Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefenseEventCollector/InfobloxBloxOneThreatDefenseEventCollector.py:79]:
```suggestion
    results: CommandResults | str = None
```

Optional[X] is equivalent to X | None (or Union[X, None]).

File: Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefenseEventCollector/InfobloxBloxOneThreatDefenseEventCollector.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        else:
+            raise NotImplementedError(f'command {command} is not implemented.')
+
+        if results:
+            return_results(results)
+    except Exception as e:
+        auth_error = isinstance(e, DemistoException) and getattr(e, 'res') is not None\
+            and e.res.status_code == 401  # pylint: disable=E1101
+        if auth_error:
+            error_msg = 'authentication error'
```
Comment (bziser) [inline Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefenseEventCollector/InfobloxBloxOneThreatDefenseEventCollector.py:96]: The error of the api is not clear?

**Review by bziser — COMMENTED** [review]
Good job

File: Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefenseEventCollector/InfobloxBloxOneThreatDefenseEventCollector.py (hunk 2 above)
Comment (esharf) [reply Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefenseEventCollector/InfobloxBloxOneThreatDefenseEventCollector.py:32]: the current_ts is used for the `to` (AKA `t1`) argument as a part of the API itself (and not just for the next run)

File: Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefenseEventCollector/InfobloxBloxOneThreatDefenseEventCollector.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+def get_events_command(client: BloxOneTDEventCollectorClient, args: dict):
+    events = client.fetch_events(
+        args['from'], args['to'],
+        min(arg_to_number(args.get('limit')) or 1000, 10000),
+        arg_to_number(args.get('offset')) or 0
+    )
+    if argToBoolean(args.get('should_push_events', False)):
+        send_events_to_xsiam(events, VENDOR, PRODUCT)
+
+    return CommandResults(outputs=events, outputs_prefix="test_get_events")
```
Comment (DeanArbel) [inline Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefenseEventCollector/InfobloxBloxOneThreatDefenseEventCollector.py:53]: `outputs_prefix="test_get_events"` is not good. Please change it to someone that matches our standards, and document it in the readme.

File: Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefenseEventCollector/README.md [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+| --- | --- | --- |
+| should_push_events | Set this argument to True in order to create events, otherwise the command will only display them. Possible values are: True, False. Default is False. | Required | 
+| from | Timestamp indicating when to start fetching events. | Required | 
+| to | Timestamp indicating when to stop fetching events. | Required | 
+| limit | Maximum number of events to fetch. Default is 1000. | Required | 
+| offset | offset of the events. | Required | 
+
+#### Context Output
+
+There is no context output for this command.
```
Comment (DeanArbel) [inline Packs/InfobloxBloxOne/Integrations/InfobloxBloxOneThreatDefenseEventCollector/README.md:45]: Please update this with a true example.

**Review by DeanArbel — APPROVED** [review]
Looks good, I have just one small note.

**Pull request merged**